src/nvme/mi-mctp.c

```diff
@@ -220,6 +226,200 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
 	return true;
 }
 
+static int nvme_mi_mctp_aem_fd(struct nvme_mi_ep *ep)
+{
+	struct nvme_mi_transport_mctp *mctp;
+
+	if (ep->transport != &nvme_mi_transport_mctp) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (!ep->transport_data) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	mctp = ep->transport_data;
+	return mctp->sd_aem;
+}
+
+static int nvme_mi_mctp_aem_purge(struct nvme_mi_ep *ep)
+{
+	char buffer[1024];
+	ssize_t bytes_read;
+
+	struct nvme_mi_transport_mctp *mctp = ep->transport_data;
+
+	// Set the socket to non-blocking mode
+	int flags = fcntl(mctp->sd_aem, F_GETFL, 0);
+
+	fcntl(mctp->sd_aem, F_SETFL, flags | O_NONBLOCK);
+
+	// Read until there is no more data
+	while ((bytes_read = recv(mctp->sd_aem, buffer, sizeof(buffer), 0)) > 0)
+		;
+
+	// Check for errors other than EAGAIN or EWOULDBLOCK
+	if (bytes_read < 0 && errno != EAGAIN && errno != EWOULDBLOCK)
+		perror("recv");
+
+	// Restore the socket to blocking mode
+	fcntl(mctp->sd_aem, F_SETFL, flags);
+
+	return 0;
+}
+
+static int nvme_mi_mctp_aem_read(struct nvme_mi_ep *ep,
+			       struct nvme_mi_resp *resp)
+{
+	ssize_t len, resp_len, resp_hdr_len, resp_data_len;
+	struct sockaddr_mctp src_addr = { 0 };
+	struct nvme_mi_transport_mctp *mctp;
+	struct iovec resp_iov[1];
+	struct msghdr resp_msg;
+	int rc, errno_save, timeout;
+	struct pollfd pollfds[1];
+	__le32 mic;
+
+	if (ep->transport != &nvme_mi_transport_mctp) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	/* we need enough space for at least a generic (/error) response */
+	if (resp->hdr_len < sizeof(struct nvme_mi_msg_hdr)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	mctp = ep->transport_data;
+
+	resp_len = resp->hdr_len + resp->data_len + sizeof(mic);
+	if (resp_len > mctp->resp_buf_aem_size) {
+		void *tmp = realloc(mctp->resp_buf_aem, resp_len);
+
+		if (!tmp) {
+			errno_save = errno;
+			nvme_msg(ep->root, LOG_ERR,
+				 "Failure allocating response buffer: %m\n");
+			errno = errno_save;
+			rc = -1;
+			goto out;
+		}
+		mctp->resp_buf_aem = tmp;
+		mctp->resp_buf_aem_size = resp_len;
+	}
+
+	/* offset by one: the MCTP message type is excluded from the buffer */
+	resp_iov[0].iov_base = mctp->resp_buf_aem + 1;
+	resp_iov[0].iov_len = resp_len - 1;
+
+	memset(&resp_msg, 0, sizeof(resp_msg));
+	resp_msg.msg_iov = resp_iov;
+	resp_msg.msg_iovlen = 1;
+	resp_msg.msg_name = &src_addr;
+	resp_msg.msg_namelen = sizeof(src_addr);
+
+	pollfds[0].fd = mctp->sd_aem;
+	pollfds[0].events = POLLIN;
+	timeout = 1;
+retry:
+	rc = ops.poll(pollfds, 1, timeout);
```

Comment:
No need to poll on the AEM socket, since we expect the caller to have already done so - just handle the `EAGAIN` on the recvmsg.